.github/workflows/ci.yaml

```diff
@@ -10,159 +16,270 @@ jobs:
     steps:
       - name: Checkout
         uses: actions/checkout@v3
-      - name: Install dependencies
+      - name: Install tox
+        # TODO: Consider replacing with custom image on self-hosted runner OR pinning version
         run: python3 -m pip install tox
       - name: Run linters
-        run: tox -e lint
+        run: tox run -e lint
 
   unit-test:
     name: Unit tests
     runs-on: ubuntu-latest
     steps:
       - name: Checkout
         uses: actions/checkout@v3
-      - name: Install dependencies
-        run: python -m pip install tox
+      - name: Install tox
+        # TODO: Consider replacing with custom image on self-hosted runner OR pinning version
+        run: python3 -m pip install tox
       - name: Run tests
-        run: tox -e unit
+        run: tox run -e unit
 
-  integration-test-lxd-ha:
-    strategy:
-      matrix:
-        series: [focal, jammy]
-    name: Integration tests for HA (lxd)
-    needs:
-      - lint
-      - unit-test
+  get-build-matrix:
+    name: Create build matrix
     runs-on: ubuntu-latest
     steps:
       - name: Checkout
         uses: actions/checkout@v3
-      - name: Setup operator environment
-        uses: charmed-kubernetes/actions-operator@main
-        with:
-          provider: lxd
-          bootstrap-options: "--agent-version 2.9.29"
-      - name: Run integration HA tests
-        run: tox -e integration-ha -- --series=${{ matrix.series }}
+      - name: Install tox
+        # TODO: Consider replacing with custom image on self-hosted runner OR pinning version
+        run: python3 -m pip install tox
+      - name: Create build matrix from charmcraft.yaml file(s)
+        id: create-matrix
+        run: tox run -e create-build-matrix
+    outputs:
+      build-matrix: ${{ steps.create-matrix.outputs.build_matrix }}
 
-  integration-test-lxd-db-router:
+  build:
     strategy:
       matrix:
-        series: [focal, jammy]
-    name: Integration tests for db-router relation (lxd)
+        charm: ${{ fromJSON(needs.get-build-matrix.outputs.build-matrix) }}
+    name: Build ${{ matrix.charm.name }} charm / ${{ matrix.charm.series }}
     needs:
-      - lint
-      - unit-test
+      - get-build-matrix
     runs-on: ubuntu-latest
     steps:
       - name: Checkout
         uses: actions/checkout@v3
-      - name: Setup operator environment
-        uses: charmed-kubernetes/actions-operator@main
+      - name: Setup environment
+        # TODO: Replace with custom image on self-hosted runner
+        run: |
+          # Copied from https://github.com/charmed-kubernetes/actions-operator/blob/96fb0b07eb675f74cf1796be812bc7e67a0d62fc/src/bootstrap/index.ts#L151
+          sudo adduser $USER lxd
+          newgrp lxd
+          sudo lxd waitready
+          sudo lxd init --auto
+          sudo iptables -F FORWARD
+          sudo iptables -P FORWARD ACCEPT
+          sudo chmod a+wr /var/snap/lxd/common/lxd/unix.socket
+          sudo snap install charmcraft --classic
```

Comment:
Did you extract this for the speed reason only? Or smth conceptually broken inside charmed-kubernetes/actions-operator@main?